File: Dynamic_Programming/3122. Minimum Number of Operations to Satisfy Conditions.cpp
/*
3122. Minimum Number of Operations to Satisfy Conditions
*/

// TC:O(m*n) SC:O(10*n)
class Solution {
    int dp[1005][10]; // 走到某一column時 如果想將前面數字改成0-9最少需要多少操作次數
public:
    int minimumOperations(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        for(int i = 0; i < 1005; i++){
            for(int j = 0; j < 10; j++){
                dp[i][j] = INT_MAX/2;
            }
        }
        for(int j = 0; j < n; j++){
            vector<int> count(10);
            for(int i = 0; i < m; i++){
                count[grid[i][j]] += 1;
            }
            
            if(j == 0){
                for(int a = 0; a < 10; a++){
                    dp[j][a] = m - count[a];
                }
            }
            else {
                for(int a = 0; a < 10; a++){
                    for(int b = 0; b < 10; b++){
                        if(a != b){
                            dp[j][a] = min(dp[j][a], m - count[a] + (j-1 >= 0 ? dp[j-1][b] : 0));
                        }
                    }
                }
            }
            
        }
        
        int res = INT_MAX;
        for(int i = 0; i < 10; i++){
            res = min(res, dp[n-1][i]);
        }
        
        return res;
        
    }
};

/*
可任意改數字 變其他正數
目標
grid[i][j] == grid[i + 1][j] 相同
grid[i][j] != grid[i][j + 1] 不同
dp[1005][10]; // 走到某一column時 如果想將前面數字改成0-9最少需要多少操作次數
*/

File: Dynamic_Programming/3122. Minimum Number of Operations to Satisfy Conditions.py
# TC:O(n^2) SC:O(n * k) k = 10
class Solution:
    def minimumOperations(self, grid: List[List[int]]) -> int:
        '''
        需要讓每個位置 i,j = i+1,j, != i, j+1 問說最少需要改幾次
        f[j][k] = f[j+1][k'] k != k
        '''
        m, n = len(grid), len(grid[0])
        # 先預處理每一行的元素個數from 0 - 9
        pre = [[0] * 10 for _ in range(n)]
        for i in range(m):
            for j in range(n):
                pre[j][grid[i][j]] += 1
        '''
        遞歸
        @cache
        def dfs(i:int, prev:int) -> int:
            if i < 0:
                return 0
            # 判斷這次要選啥然後所需的cost 去取最小的 並且不能跟上次的一樣
            ans = inf
            for k in range(10):
                if k != prev:
                    ans = min(ans, dfs(i-1, k) + m - pre[i][k])
            return ans
        return dfs(n-1, -1)
        '''
        f = [[inf] * 10 for _ in range(n)]
        for j in range(n-1, -1, -1):
            # 計算上一行最小的元素與次小的
            mn, mn2, num = inf, inf, -1
            if j + 1 < n:
                for k in range(10):
                    if f[j+1][k] <= mn:
                        mn, mn2, num = f[j+1][k], mn, k
                    elif f[j+1][k] < mn2: mn2 = f[j+1][k]
            if num == -1:
                for k in range(10):
                    f[j][k] = m - pre[j][k]
            else:
                for k in range(10):
                    if k != num:
                        f[j][k] = m - pre[j][k] + mn
                    else:
                        f[j][k] = m - pre[j][k] + mn2
        return min(f[0][k] for k in range(10))
        
        

